### PR TITLE
[5.1] Allow relationships in factory builder

### PR DESCRIPTION
**Pull request by russweas**
(I'm pretty sure I didn't do this perfectly but I hope that someone who is better than me at this gets the idea.) :)
Basically I can put something like this in my FactoryBuilder.php file:

'user_id' => App\User::class

and it will take the id of the user and set it as the attribute. I like this better than 'factory:User' because it will allow me to click through to the User model. If there's any way to make this cleaner or faster just tell me and I'll fix it ASAP! Especially the try catch because I feel like it's kind of hacky. Also comment if you think I should export this to its own function

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 1]
```diff
@@ -127,6 +127,16 @@ protected function makeInstance(array $attributes = [])
             }
 
             $definition = call_user_func($this->definitions[$this->class][$this->name], $this->faker);
+            
+            foreach($definition as $key => $attribute)
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: cs

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 2]
```diff
@@ -127,6 +127,16 @@ protected function makeInstance(array $attributes = [])
...
+            {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: cs

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 3]
```diff
@@ -127,6 +127,16 @@ protected function makeInstance(array $attributes = [])
...
+                try {
+                    $class = app($attribute);
+                    if($class instanceof Model)
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: cs

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 4]
```diff
@@ -127,6 +127,16 @@ protected function makeInstance(array $attributes = [])
...
+                        $definition[$key] = $class::orderByRaw("RAND()")->get()->id;
+                } 
+                catch(Exception $e) {   }
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: cs

**Comment by rweas** [comment]
@GrahamCampbell Sorry I don't know what cs means :/

**Comment by GrahamCampbell** [comment]
Coding style. We follow PSR-2 with a few extra things.

**Comment by GrahamCampbell** [comment]
Few extra things are the phpdoc style and things like we don't want 6 newlines before closing a class, for example, which PSR-2 neither allows or disallows.

**Comment by rweas** [comment]
Oh yeah. Not used to it yet. I'll fix that and then if everything is good squash the commits.

**Comment by GrahamCampbell** [comment]
Thanks. :)

**Comment by rweas** [comment]
I can't currently push my changes to githib but is the try catch block ok?

**Comment by GrahamCampbell** [comment]
No, needs to be like this:

```php
try {
    // code
} catch (Exception $e) {
    //
}
```

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 5]
```diff
@@ -127,6 +127,14 @@ protected function makeInstance(array $attributes = [])
             }
 
             $definition = call_user_func($this->definitions[$this->class][$this->name], $this->faker);
+            
+                try {
+                    $class = app($attribute);
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: You can't use the app helper in the core.

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 6]
```diff
@@ -127,6 +127,14 @@ protected function makeInstance(array $attributes = [])
...
+                    if($class instanceof Model)
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: you ned a space after `if`

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: You muse include braces.

**Comment by rweas** [comment]
Cool! I'll get those things fixed ASAP.

**Comment by rweas** [comment]
I decided to just use php's built in reflection since I don't know the Laravel preffered way to do it without app(). @GrahamCampbell just double checking is that OK?

**Comment by rweas** [comment]
Besides that it works perfectly in a test project and a real project.

**Comment by rkgrep** [comment]
You can't use `orderByRaw('RAND()')` as it won't work on Postgres - the function is called `random()` there.

**Comment by rweas** [comment]
@rkgrep I should add a note to the docs saying this for now. Later I want to create a PR to add some sort of Model::random() that will do this automatically. It should work for both MySQL and Postgres. I probably wont get to it for some time between two weeks and a month though.

**Comment by rkgrep** [comment]
Such things cannot be commented in doc like *Don't use on PostgreSQL*
Also I don't like using reflections here.

I think you'd better improve your factory definition

````php
$factory->define(App\User::class, function ($faker) {
    function getRandomGroup()
    {
        return object_get(App\Group::orderByRaw('RAND()')->first(), 'id', null);
    }
    return [
        'name' => $faker->name,
        'email' => $faker->email,
        'group_id' => getRandomGroup(),
        'password' => str_random(10),
        'remember_token' => str_random(10),
    ];
});
````

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 7]
```diff
@@ -127,6 +129,16 @@ protected function makeInstance(array $attributes = [])
...
 
             $definition = call_user_func($this->definitions[$this->class][$this->name], $this->faker);
+            
+            foreach ($definition as $key => $attribute) {
+                try {
+                    $class = new ReflectionClass($attribute);
+                    if ($class instanceof Model) {
+                        $definition[$key] = $class::orderByRaw('RAND()')->first()->id;
+                    }
+                } catch (\ReflectionException $e) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: no slash

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 8]
```diff
@@ -4,6 +4,8 @@
 
 use Faker\Factory as Faker;
 use InvalidArgumentException;
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: sort these by length please

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 9]
```diff
@@ -127,6 +129,16 @@ protected function makeInstance(array $attributes = [])
             }
 
             $definition = call_user_func($this->definitions[$this->class][$this->name], $this->faker);
+            
+            foreach ($definition as $key => $attribute) {
+                try {
+                    $class = new ReflectionClass($attribute);
+                    if ($class instanceof Model) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: surely this is always false?

**Comment by rkgrep** [comment]
@GrahamCampbell is there any sense to continue fixing CS?

As I mentioned - this fix will completely break ability to generate polymorphic models - as it will try to generate ids instead of polymorphic_type field.

**Comment by taylorotwell** [comment]
Not a big fan of this implementation.

**Comment by rweas** [comment]
Was on a vacation last week and totally forgot about this until now. @taylorotwell is there a better way you would recommend me to implement it? I would definitely be up for it!

**Comment by masterpowers** [comment]
Hi im getting this error when i try to use the code above
 Cannot redeclare getRandomSponsor()

**Pull request closed without merging**